### PR TITLE
Fix batch norm training op on CPU

### PR DESCRIPTION
**Pull request by pranav-prakash**
**Description**:  Add support for training-forward-mode of BatchNorm on CPU EP, and implement BN Gradient on CPU EP

Fixes the issue described in #6087

**Motivation and Context**

As mentioned in the above issue, currently batch norm is not implemented for training. This PR makes the following changes:

* Blacklist the mean/variance tensors from having their gradients calculated. These are calculated directly from the batch during training and are not updated via backprop
* The onnx spec for BatchNorm currently doesn't provide an attribute to distinguish between training/inference mode (since batch norm has different behavior in both cases). onnx/onnx#1042 While we can assume the presence of the optional outputs indicates we're doing training, it's technically valid for a serialized model for inference to contain those as well. The CUDA kernel nonetheless uses this as an indicator of training, and so do we.
* The CPU implementation of the above was cribbed from caffe2 (as it appeared that the existing inference-only implementation was also taken from there). They do something weird where instead of outputting saved_variance they output inv_std_dev. Apparently this is for ease of interoperability with cuDNN, but it completely breaks the ONNX spec. Nonetheless, I've chosen to do the same because the existing cuda kernel for batchnormgrad also relies on it actually being inv_std_dev.
* Added a CPU implementation for batchnormgrad. Again to match behavior with the cuda version we too completely break the spec and assume that the saved_var is actually inv_std_dev.

As a sidenote, I wonder if the flaky CUDA tests for the batch norm can also be resolved by https://github.com/microsoft/onnxruntime/commit/d01006fc222799f879c8ce70edd58e97f53b7767

### CODE REVIEW

**Comment by pranav-prakash** [comment]
As a sidenote, I wonder if the flaky CUDA tests for the batch norm can also be resolved by https://github.com/microsoft/onnxruntime/commit/d01006fc222799f879c8ce70edd58e97f53b7767

**Comment by pranav-prakash** [comment]
@SherlockNoMad @mrry Could either of you review the PR?

**Comment by neginraoof** [comment]
Hey @pranav-prakash 
Can you also review this PR to fix onnx spec? https://github.com/onnx/onnx/pull/3333

**Comment by pranav-prakash** [comment]
@neginraoof I'm not a msft employee, so I don't have any more power than you do with regard to code review

**Comment by pranav-prakash** [comment]
Given that the training mode spec for BN is only fully fleshed out for opset 14, do we still need to support the case of training for non-spatial BN? Removing this codepath would simplify things greatly (we could still maintain correctness by guarding training on opset 14 with `ORT_ENFORCE`).

**Comment by SherlockNoMad** [comment]
> Given that the training mode spec for BN is only fully fleshed out for opset 14, do we still need to support the case of training for non-spatial BN (which could happen with opset < 14)? Removing this codepath would simplify things greatly (we could still maintain correctness by guarding training on opset 14 with `ORT_ENFORCE`).

yes. Since BN training mode is only officially added since opset 14, let's ease the burden of supporting the non-spatial mode for training. :)

**Comment by SherlockNoMad** [comment]
Hi @pranav-prakash, thanks a lot for your contribution. 
Are you planning to update the PR to deprecate the support for non-spatial mode in training? I can help you with review and merging the PR.

**Comment by SherlockNoMad** [comment]
Hi @pranav-prakash, I am also wondering what's your use scenario of ORT training? Which company/product are you working on?

**Comment by pranav-prakash** [comment]
@SherlockNoMad
>update the PR to deprecate the support for non-spatial mode in training
Yes I started the PR for this, but then saw that the spec was also updated to remove the `saved_mean` and `saved_var` outputs. With these two removed, it seems that you would have to recompute the mean/var in the gradient op. I had asked about the motivation for this change on the associated [PR](https://github.com/onnx/onnx/pull/3379#discussion_r606490791), and it seems that the intent was to allow "backends [to] transform the graph/node into a custom-op of their choice (for backward-propagation)."

In terms of ORT though, does this mean that we would define our own variant of the BatchNorm schema that includes outputs for `saved_mean` and `saved_inv_std` and use a graph transform pass to change nodes to this? If not, is there another way to avoid the recomputation of `batch_mean`/`batch_var` in the backward pass?

>I am also wondering what's your use scenario of ORT training? Which company/product are you working on?
I'm associated with UC Berkeley's [architecture research group](https://bar.eecs.berkeley.edu/index.html), and we're [working on](https://github.com/pranav-prakash/onnxruntime-riscv/) an ORT EP for our risc-v ml acceleator ([Gemmini](https://arxiv.org/abs/1911.09925)). In terms of use-cases we're primarily interested in training convolutional neural networks (in both bfloat16 and fp32), e.g. resnet or mobilenet.

**Comment by SherlockNoMad** [comment]
Thanks a lot for the detailed introduction! We are really happy to see external contribution to ORT Training!!!

As the for the BatchNorm problem, the plan is to write a custom op (say, BatchNormInternal)  that outputs the saved_mean and save_inv_std. We will substitute BatchNorm with BatchNormInternal before building the training graph. The BatchNormGrad op can still assume that O(3) and O(4) are present for speed up the computation. 

The rational behind the onnx spec update is that, the "save_inv_std" is an internal implementation detail, (other framework may use save_inv_var instead), so it's better to leave it out of the spec.

**Comment by SherlockNoMad** [comment]
Just curious, how far have your reached for training convolutional neural networks e.g. resnet or mobilenet? Are they working yet? We are also exploring federated learning on mobile/edge device. Would be nice to have a colab if our plans aligns well.

**Comment by pranav-prakash** [comment]
@SherlockNoMad
>the plan is to write a custom op (say, BatchNormInternal) that outputs the saved_mean and save_inv_std
I see – so for this PR would you like me to create the schema for such a `BatchNormTrainingInternal` and move the training-mode calculations there? Or did you still want to have an implementation for `training_mode = true` in the `BatchNorm` kernel just for completeness (albeit one that will never be used if the op gets replaced with `BatchNormTrainingInternal` before training).

>how far have your reached for training convolutional neural networks
We just recently got a [trainer](https://github.com/pranav-prakash/onnxruntime-riscv/tree/2021-04-02/systolic_runner/imagenet_trainer) for resnet50 working, although we haven't yet verified end-to-end correctness since fpga simulation of Gemmini is much too slow for training resnet from scratch (we could likely compare results from a few dozen iterations against the CPU EP though).

 >We are also exploring federated learning on mobile/edge device. Would be nice to have a colab if our plans aligns well.
Yeah federated learning and training at the edge are exactly the scenarios we envisioned Gemmini would be a good fit for. We'd love to discuss our roadmap and see if there's any potential for collaboration here; feel free to email us at `{pranavprakash,hngenc}@berkeley <dot> edu`.

**Comment by SherlockNoMad** [comment]
BatchNormTrainingInternal and BatchNorm can share the same kernel. Most of the code should be same, except for the handling of output 3 and 4. In the kernel, we can check if the ctx->Output(3, shape) and ctx->Output(4, shape) returns the nullptr. 

If you have bandwidth, you can add the BatchNormalizationTraining Schema to training_op_defs.cc. 

Also, a sample of the replacement code can be found in concat_replacement.cc. This replaces Concat with ConcatTraining.

**Comment by SherlockNoMad** [comment]
@pranav-prakash. Great to know that you got the resnet50 working... 
Actually I am a bit surprised that you didn't find too many missing gradients.... 

AFAIK, gradient is missing for LRN, Sum, GlobalMaxPool op... GlobalAveragePool's gradient builder can also need a fix...

**Comment by pranav-prakash** [comment]
>BatchNormTrainingInternal and BatchNorm can share the same kernel

SG, I'll update the PR accordingly. Not sure I'll have the bandwidth to add the `BatchNormalizationTraining` op as well, but that should be an easy subsequent PR.

>bit surprised that you didn't find too many missing gradients

We converted `Sum` into `Add` before training; iirc both the resnet50 structure from the model-zoo and exported pytorch use `MaxPool` instead of `GlobalMaxPool`. The pytorch model does use `GlobalAveragePool` but we didn't seem to run into any issues with the gradient builder on that.

**Comment by pranav-prakash** [comment]
@SherlockNoMad
I've updated the BN kernel to support the opset-14 case. I don't think the onnx submodule has been bumped to the latest commit though, so that will need to be done before this can be merged.

Since the batch norm grad cannot be implemented until the schema & graph transform for internal `BatchNormalizationTraining` is added, I've reverted that for now. For reference, those can be found at
https://github.com/microsoft/onnxruntime/blob/8892ee4b6d343109699ab292e66c2c7a5e41925a/orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.h
https://github.com/microsoft/onnxruntime/blob/8892ee4b6d343109699ab292e66c2c7a5e41925a/orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc

**Comment by SherlockNoMad** [comment]
https://github.com/microsoft/onnxruntime/pull/7177 
is in progress.

**Comment by pranav-prakash** [comment]
@SherlockNoMad
Now that the onnx submodule has been bumped to opset 14, I updated this PR to enable the opset 14 BN. Also fixed the previous CI failure.

**Comment by SherlockNoMad** [comment]
thanks a lot @pranav-prakash 
I will take a look and schedule the CI for you.

**Comment by pranav-prakash** [comment]
@SherlockNoMad Is the failure on the TensorRT & OpenVINO pipelines due to something specific to those backends? I see
```
[E:onnxruntime:Default, provider_test_utils.cc:597 onnxruntime::test::OpTester::ExecuteModel] Initialize failed with status: Exception during initialization: invalid vector subscript
```
which isn't too informative, and none of the CPU/GPU pipelines seem to be affected. I do see some tests in  in `batch_norm_op_test` exclude `kTensorrtExecutionProvider` and `kOpenVINOExecutionProvider` but I don't want to blindly copy them without understanding where the issue lies.

**Comment by SherlockNoMad** [comment]
see line 44 in batch_norm_op_test.cc

```
  std::unordered_set<std::string> excluded_eps = {kTensorrtExecutionProvider};
  if (spatial_mode == 0) {
    excluded_eps.insert(kOpenVINOExecutionProvider);
  }
```

I think it's correct to set them in excluded_eps .

**Review by SherlockNoMad — APPROVED** [review]
Thanks a lot for this PR. 
I think it's ready to be merge, after the fix for the TRT and openVino fix.

**Comment by pranav-prakash** [comment]
@SherlockNoMad fixed.

One other question I had was about the old cuda-test for `ForwardTrainingTestWithSavedOutputsOpset9`. The old test had
```
  test.AddOutput<float>("running_mean", channel_dims, {-0.1754f, 0.303106f});
  test.AddOutput<float>("saved_mean", channel_dims, {-0.306f, 0.115f});
```
That is, it had `running_mean` equal to the `saved_mean`, when according to the formula should have been equal to `momentum*mean + saved_mean*(1-momentum)`. Was the old test incorrect, or did I miss something obvious?

**Comment by SherlockNoMad** [comment]
I think the old test case was incorrect.

**Review by SherlockNoMad — DISMISSED** [review]

File: onnxruntime/core/providers/cpu/cpu_execution_provider.cc [hunk 1]
```diff
@@ -286,8 +286,10 @@ class ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOn
 class ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 9, 12, double, MatMul);
 class ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 9, 12, int32_t, MatMul);
 class ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 9, 12, int64_t, MatMul);
-class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 9, float, BatchNormalization);
-class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 9, double, BatchNormalization);
+class ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 9, 13, float, BatchNormalization);
+class ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 9, 13, double, BatchNormalization);
+class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 14, float, BatchNormalization);
```
Comment (askhade) [inline onnxruntime/core/providers/cpu/cpu_execution_provider.cc:291]: please add opset 14 registrations together with other opset 14 registrations

File: onnxruntime/core/providers/cpu/cpu_execution_provider.cc [hunk 2]
```diff
@@ -1143,9 +1145,13 @@ Status RegisterOnnxOperatorKernels(KernelRegistry& kernel_registry) {
...
                                                                             MatMul)>,
-      BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 9, float,
+      BuildKernelCreateInfo<ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 9, 13, float,
                                                                   BatchNormalization)>,
-      BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 9, double,
+      BuildKernelCreateInfo<ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 9, 13, double,
+                                                                  BatchNormalization)>,
+      BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 14, float,
+                                                                  BatchNormalization)>,
+      BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 14, double,
```
Comment (askhade) [inline onnxruntime/core/providers/cpu/cpu_execution_provider.cc:1154]: please add opset 14 registrations together with other opset 14 registrations

**Review by SherlockNoMad — APPROVED** [review]
